File: src/main/java/xyz/phanta/cbtweaker/singleblock/SingleBlockManager.java
package xyz.phanta.cbtweaker.singleblock;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.util.math.Vec2i;
import net.minecraft.util.ResourceLocation;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.ItemBlockRegistrar;
import xyz.phanta.cbtweaker.buffer.BufferGroup;
import xyz.phanta.cbtweaker.buffer.BufferType;
import xyz.phanta.cbtweaker.buffer.BufferTypeRegistry;
import xyz.phanta.cbtweaker.common.BlockMaterial;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;
import xyz.phanta.cbtweaker.gui.ScreenRegion;
import xyz.phanta.cbtweaker.gui.inventory.UiLayout;
import xyz.phanta.cbtweaker.recipe.RecipeLogic;
import xyz.phanta.cbtweaker.util.ConfigException;
import xyz.phanta.cbtweaker.util.helper.DataLoadUtils;

import javax.annotation.Nullable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

public class SingleBlockManager {

    private static final UiLayout DEFAULT_UI_LAYOUT = new UiLayout(
            CbtTextureResources.GUI_GENERIC_SMALL, new ScreenRegion(7, 16, 162, 55), true, new Vec2i(7, 83), true);

    private final Path specDir;
    private final Map<String, JsonObject> sbSpecDtoCache = new HashMap<>();
    private final Map<String, SingleBlockType<?, ?, ?>> sbTypeTable = new HashMap<>();
    private final ItemBlockRegistrar registrar;

    public SingleBlockManager(Path specDir, ItemBlockRegistrar registrar) {
        this.specDir = specDir;
        this.registrar = registrar;
    }

    public void preloadAll() {
        CbtMod.LOGGER.info("Loading single-block types...");
        try {
            Files.list(specDir).forEach(sbDir -> {
                try {
                    if (!Files.isDirectory(sbDir)) {
                        return;
                    }
                    Path specFile = sbDir.resolve("singleblock.json");
                    if (!Files.isRegularFile(specFile)) {
                        CbtMod.LOGGER.warn("Ignoring single-block directory without a singleblock.json: {}", sbDir);
                        return;
                    }

                    JsonObject specDto = DataLoadUtils.readJsonFile(specFile).getAsJsonObject();
                    BlockMaterial blockMat = specDto.has("material")
                            ? BlockMaterial.fromString(specDto.get("material").getAsString()) : BlockMaterial.METAL;
                    RecipeLogic<?, ?, ?, ?, ?> recipeLogic = DataLoadUtils.loadRecipeLogic(
                            specDto.getAsJsonObject("recipe_logic"));
                    UiLayout uiLayout = specDto.has("ui")
                            ? CbtMod.PROXY.getTemplates().getUiTemplates().resolve(specDto.get("ui"))
                            : DEFAULT_UI_LAYOUT;

                    JsonObject buffersDto = specDto.getAsJsonObject("buffers");
                    SortedMap<String, BufferGroup.Factory> bufGroupFactories = new TreeMap<>();
                    for (Map.Entry<String, JsonElement> bufGroupDtoEntry : buffersDto.entrySet()) {
                        BufferGroup.Factory factory = new BufferGroup.Factory();
                        for (Map.Entry<String, JsonElement> bufferDtoEntry
                                : bufGroupDtoEntry.getValue().getAsJsonObject().entrySet()) {
                            ResourceLocation bufTypeId = new ResourceLocation(bufferDtoEntry.getKey());
                            BufferTypeRegistry.Entry<?, ?, ?, ?> bufTypeEntry = CbtMod.PROXY.getBufferTypes()
                                    .lookUp(bufTypeId);
                            if (bufTypeEntry == null) {
                                throw new ConfigException("Unknown buffer type: " + bufTypeId);
                            }
                            JsonElement bufferListDto = bufferDtoEntry.getValue();
                            if (bufferListDto.isJsonArray()) {
                                for (JsonElement bufferDto : bufferListDto.getAsJsonArray()) {
                                    loadBufferFactory(
                                            factory, bufTypeEntry.getBufferType(), bufferDto.getAsJsonObject());
                                }
                            } else {
                                loadBufferFactory(
                                        factory, bufTypeEntry.getBufferType(), bufferListDto.getAsJsonObject());
                            }
                        }
                        bufGroupFactories.put(bufGroupDtoEntry.getKey(), factory);
                    }

                    SingleBlockType<?, ?, ?> sbType = new SingleBlockType<>(
                            sbDir.getFileName().toString(), blockMat, bufGroupFactories, sbDir, recipeLogic, uiLayout);
                    sbSpecDtoCache.put(sbType.getId(), specDto);
                    sbTypeTable.put(sbType.getId(), sbType);
                    registrar.addBlockRegistration("sb_" + sbType.getId(), new SingleBlockMachineBlock(sbType));
                    CbtMod.LOGGER.debug("Loaded single-block: {}", sbType.getId());
                } catch (Exception e) {
                    throw new IllegalStateException("Failed to load single-block specification: " + sbDir, e);
                }
            });
        } catch (Exception e) {
            throw new IllegalStateException("Failed to load single-block types!", e);
        }
        CbtMod.LOGGER.info("Finished loading single-block specifications.");
    }

    private static <B> void loadBufferFactory(BufferGroup.Factory bufGroupFactory, BufferType<B, ?, ?, ?> bufType,
                                              JsonObject bufferDto) {
        bufGroupFactory.addFactory(bufType, bufType.loadBufferFactory(bufferDto));
    }

    public void loadAll() {
        CbtMod.LOGGER.info("Loading single-block data...");
        for (SingleBlockType<?, ?, ?> sbType : sbTypeTable.values()) {
            try {
                lateInitSingleBlockType(sbType, sbSpecDtoCache.get(sbType.getId()).getAsJsonObject("recipe_logic"));
                CbtMod.LOGGER.debug("Initialized single-block: {}", sbType.getId());
            } catch (Exception e) {
                throw new IllegalStateException("Failed to load data for single-block: " + sbType.getId(), e);
            }
        }
        sbSpecDtoCache.clear();

        CbtMod.LOGGER.info("Loading single-block recipes...");
        for (SingleBlockType<?, ?, ?> sbType : sbTypeTable.values()) {
            loadRecipesForSingleBlock(sbType);
        }
        CbtMod.LOGGER.info("Finished loading single-block data.");
    }

    private static <R, C, D> void lateInitSingleBlockType(SingleBlockType<R, C, D> sbType,
                                                          JsonObject recipeLogicDto) {
        RecipeLogic<R, C, D, ?, ?> recipeLogic = sbType.getRecipeLogic();
        C recipeConfig = recipeLogic.loadConfig(recipeLogicDto);
        sbType.init(recipeConfig, recipeLogic.createEmptyDatabase(recipeConfig));
    }

    private <R, C, D> void loadRecipesForSingleBlock(SingleBlockType<R, C, D> sbType) {
        try {
            DataLoadUtils.loadRecipes(sbType.getId(), sbType.getRecipeDirectory(), "singleblock.json",
                    sbType.getRecipeLogic(), sbType.getRecipeConfig(), sbType.getRecipeDatabase());
        } catch (Exception e) {
            throw new IllegalStateException("Failed to load recipes for single-block: " + sbType.getId());
        }
    }

    public Collection<SingleBlockType<?, ?, ?>> getAll() {
        return sbTypeTable.values();
    }

    @Nullable
    public SingleBlockType<?, ?, ?> lookUp(String id) {
        return sbTypeTable.get(id);
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/singleblock/SingleBlockManager.kt
package st.evening.mc.cbtweaker.singleblock

import it.unimi.dsi.fastutil.objects.Object2ObjectRBTreeMap
import it.unimi.dsi.fastutil.objects.Object2ObjectSortedMaps
import net.minecraft.util.ResourceLocation
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.buffer.BufferGroup
import st.evening.mc.cbtweaker.common.BlockConfig
import st.evening.mc.cbtweaker.gui.inventory.WindowConfig
import st.evening.mc.cbtweaker.util.DataGenHelper
import st.evening.mc.cbtweaker.util.forEachFile
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.TypedJsonParser
import st.evening.mc.prelude.api.data.tjson.forEachObject
import st.evening.mc.prelude.api.data.tjson.useAny
import st.evening.mc.prelude.api.data.tjson.useObject
import st.evening.mc.prelude.api.data.tjson.useObjectValue
import st.evening.mc.prelude.api.data.tjson.useStringValue
import st.evening.mc.prelude.api.registration.ModRegistrar
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.readText

class SingleBlockManager(private val specDir: Path, private val reg: ModRegistrar) : Iterable<SingleBlockType<*>> {
    private val sbSpecDtoCache: MutableMap<String, TJson.Object> = mutableMapOf()
    private val sbTypeTable: MutableMap<String, SingleBlockType<*>> = mutableMapOf()

    operator fun get(id: String): SingleBlockType<*>? = sbTypeTable[id]

    override fun iterator(): Iterator<SingleBlockType<*>> = sbTypeTable.values.iterator()

    fun preloadAll() {
        CbTweaker.logger.info("Loading single-block types...")
        try {
            specDir.forEachFile { specFile ->
                try {
                    if (!Files.isRegularFile(specFile)) return@forEachFile
                    val specFileName = specFile.fileName.toString()
                    if (!specFileName.endsWith(".tjson")) {
                        CbTweaker.logger.warn(
                            "Ignoring non-TJSON file in single-block specification directory: {}",
                            specFileName
                        )
                        return@forEachFile
                    }
                    val sbId = specFileName.dropLast(6)
                    if (sbId.isBlank()) {
                        throw SerializationException("Empty single-block ID!")
                    }

                    val specDto = TypedJsonParser.parseObject(specFile.readText())
                    val sbType = JsonPath.atRoot {
                        SingleBlockType(
                            reg,
                            sbId,
                            specDto.useObject("block") { BlockConfig.load(it) } ?: BlockConfig.DEFAULT,
                            specDto.useObject("buffers") { bufGroupsDto ->
                                val bufGroupFactories = Object2ObjectRBTreeMap<String, BufferGroup.Factory>()
                                bufGroupsDto.forEachObject { bufGroupId, bufGroupDto ->
                                    val factory = BufferGroup.Factory()
                                    bufGroupDto.forEachObject { bufTypeId, buffersDto ->
                                        val bufType = (CbTweaker.defns.bufferTypes[ResourceLocation(bufTypeId)]
                                            ?: throw SerializationException.withPath("Unknown buffer type: $bufTypeId"))
                                            .bufferType
                                        buffersDto.forEachObject { name, bufferDto ->
                                            factory.loadFactory(bufType, name, bufferDto)
                                        }
                                    }
                                    bufGroupFactories[bufGroupId] = factory
                                }
                                return@useObject bufGroupFactories
                            } ?: Object2ObjectSortedMaps.emptyMap(),
                            specDto.useObjectValue("behaviour") { behaviourDto ->
                                behaviourDto.useStringValue("type") {
                                    CbTweaker.defns.machineBehaviours[ResourceLocation(it)]
                                        ?: throw SerializationException.withPath("Unknown machine behaviour: $it")
                                }
                            },
                            specDto.useAny("ui") {
                                CbTweaker.defns.templates.windowTemplates.resolve(it)
                            } ?: WindowConfig.GENERIC_SMALL
                        )
                    }
                    sbSpecDtoCache[sbId] = specDto
                    sbTypeTable[sbId] = sbType
                    CbTweaker.logger.debug("Loaded single-block: {}", sbId)
                } catch (e: Exception) {
                    throw IllegalStateException("Failed to load single-block specification: $specFile", e)
                }
            }
        } catch (e: Exception) {
            throw IllegalStateException("Failed to load single-block types!", e)
        }
        CbTweaker.logger.info("Finished loading single-block specifications.")
    }

    fun dataGenBlockModels(resourceDir: Path) {
        if (sbTypeTable.isEmpty()) return
        val cbtDir = resourceDir.resolve("cbtweaker")
        val blockStateDir = Files.createDirectories(cbtDir.resolve("blockstates"))
        val itemModelDir = Files.createDirectories(cbtDir.resolve("models/item"))
        sbTypeTable.keys.forEach { id ->
            val bsFile = blockStateDir.resolve("sb_$id.json")
            if (!Files.exists(bsFile)) {
                CbTweaker.logger.info("Generating single-block machine block state mapping: ${bsFile.fileName}")
                DataGenHelper.writeToFile(bsFile, DataGenHelper.machineBlockState)
            }
            val imFile = itemModelDir.resolve("sb_$id.json")
            if (!Files.exists(imFile)) {
                CbTweaker.logger.info("Generating single-block machine item model: ${imFile.fileName}")
                DataGenHelper.writeToFile(imFile, DataGenHelper.machineItemModel)
            }
        }
    }

    fun loadAll() {
        CbTweaker.logger.info("Loading single-block data...")
        sbTypeTable.forEach { (sbId, sbType) ->
            try {
                lateInitSingleBlockType(sbType, sbSpecDtoCache.getValue(sbId))
                CbTweaker.logger.debug("Initialized single-block: {}", sbId)
            } catch (e: Exception) {
                throw IllegalStateException("Failed to load data for single-block: $sbId", e)
            }
        }
        sbSpecDtoCache.clear()
    }

    private fun <S> lateInitSingleBlockType(sbType: SingleBlockType<S>, specDto: TJson.Object) {
        JsonPath.atRoot {
            sbType.init(
                specDto.useObjectValue("behaviour") {
                    sbType.behaviour.loadStateFactory(sbType, it)
                }
            )
        }
    }
}
